### PR TITLE
Fix panic in median "AggregateState is not a scalar aggregate"

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

closes https://github.com/apache/arrow-datafusion/issues/3105

# Rationale for this change

I think median as currently implemented is unusable except for everything but the simplest of queries (where there is no intermediate repartitioning)

Median was the only aggregate that stored it state using `AggregateState::Array` while all other aggregages use `AggregateState::Scalar`

It turns out that `AggregateState::Array` doesn't support partial aggregates where partial state has to be marshalled into an `Array` and combined. Thus, `median` likely never worked for any DataFusion plan that has more than one partition (where the data is repartitioned for parallel aggregation).

# What changes are included in this PR?

I tried a few ways to implement partial state for `AggregateState::Array`  and they all got messy very quickly.

Thus, I opted for what I think is the simpler, possibly less performant, approach of using Scalars rather than Arrays. 

Median was added in https://github.com/apache/arrow-datafusion/pull/3009 by @andygrove (and sadly I think I have have suggested the array based approach without fully understanding the implications)

For any real world dataset where the performance different between approaches would be measurable, I believe the the current implementation is going to error anyways. 

This approach will also extend nicely to Decimal as well

# Are these changes tested?
Yes

# Are there any user-facing changes?
Less error

# Proposed follow on

If this approach is accepted, I propose removing the AggregateState enum and update the Accumulator trait to simplify it as a follow on PR.

cc @andygrove as he originally wrote this code in https://github.com/apache/arrow-datafusion/pull/3009 

cc @tustvold as I think he is thinking about how to improve the grouping situation overall

### CODE REVIEW

File: datafusion/core/tests/sql/aggregates.rs [hunk 1]
```diff
@@ -436,6 +436,90 @@ async fn median_test(
     Ok(())
 }
 
+#[tokio::test]
+// test case for https://github.com/apache/arrow-datafusion/issues/3105
```
Comment (alamb) [inline datafusion/core/tests/sql/aggregates.rs:440]: all of the tests added in this file panic on master

File: datafusion/physical-expr/src/aggregate/median.rs [hunk 2]
```diff
@@ -91,157 +91,124 @@ impl AggregateExpr for Median {
...
-        self.all_values.extend_from_slice(&[x]);
-        Ok(())
-    }
+        assert_eq!(values.len(), 1);
+        let array = &values[0];
 
-    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
-        for array in states {
-            self.all_values.extend_from_slice(&[array.clone()]);
+        self.all_values.reserve(self.all_values.len() + array.len());
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/median.rs:114]:
```suggestion
        assert!(matches!(array.data_type(), DataType::List(_)));
        self.all_values.reserve(self.all_values.len() + array.len());
```
?

File: datafusion/physical-expr/src/aggregate/median.rs [hunk 3]
```diff
@@ -91,157 +91,124 @@ impl AggregateExpr for Median {
...
-            DataType::Float32 => {
-                median!(self, arrow::datatypes::Float32Type, Float32, 2_f32)
-            }
-            DataType::Float64 => {
-                median!(self, arrow::datatypes::Float64Type, Float64, 2_f64)
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        assert_eq!(states.len(), 1);
+
+        let array = &states[0];
+        for index in 0..array.len() {
```
Comment (tustvold) [inline datafusion/physical-expr/src/aggregate/median.rs:127]:
```suggestion
        assert!(matches!(array.data_type(), DataType::List(_)));
        for index in 0..array.len() {
```

**Review by tustvold — APPROVED** [review]

File: datafusion/core/tests/sqllogictests/test_files/aggregate.slt [hunk 4]
```diff
@@ -216,6 +216,70 @@ SELECT approx_median(a) FROM median_f64_nan
 ----
 NaN
 
+# median_multi
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/aggregate.slt:219]: I ported the tests to sqllogictest as much of the rest of the aggregate tests had been ported too

File: datafusion/physical-expr/src/aggregate/median.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/aggregate/median.rs:115]: For what it is worth, this assert fails -- the correct assertion is 

```rust
        assert_eq!(array.data_type(), &self.data_type);
```

Which I have fixed

**Pull request merged**